pkg/ingester/checkpoint.go

```diff
@@ -143,55 +177,108 @@ func (i *ingesterSeriesIter) Stop() {
-			streams := make([]*stream, 0, len(inst.streams))
-			inst.streamsMtx.RUnlock()
-			_ = inst.forAllStreams(func(stream *stream) error {
-				streams = append(streams, stream)
-				return nil
-			})
-
-			for _, stream := range streams {
-				stream.chunkMtx.RLock()
-				if len(stream.chunks) < 1 {
-					stream.chunkMtx.RUnlock()
-					// it's possible the stream has been flushed to storage
-					// in between starting the checkpointing process and
-					// checkpointing this stream.
-					continue
-				}
-
-				// TODO(owen-d): use a pool
-				chunks, err := toWireChunks(stream.chunks, nil)
-				stream.chunkMtx.RUnlock()
-
-				var s *Series
-				if err == nil {
-					s = &Series{
-						UserID:      inst.instanceID,
-						Fingerprint: uint64(stream.fp),
-						Labels:      client.FromLabelsToLabelAdapters(stream.labels),
-						Chunks:      chunks,
-					}
-				}
-				select {
-				case ch <- &SeriesWithErr{
-					Err:    err,
-					Series: s,
-				}:
-				case <-i.done:
-					return
-				}
-			}
+func (i *ingesterSeriesIter) Iter() *streamIterator {
+	return newStreamsIterator(i.ing)
+}
+
+type streamInstance struct {
+	id      string
+	streams []*stream
+}
+
+type streamIterator struct {
+	instances []streamInstance
+
+	current Series
+	buffer  []chunkWithBuffer
+	err     error
+}
+
+// newStreamsIterator returns a new stream iterators that iterates over one instance at a time, then
+// each stream per instances.
+func newStreamsIterator(ing ingesterInstances) *streamIterator {
+	instances := ing.getInstances()
+	streamInstances := make([]streamInstance, len(instances))
+	for i, inst := range ing.getInstances() {
+		inst.streamsMtx.RLock()
+		streams := make([]*stream, 0, len(inst.streams))
+		inst.streamsMtx.RUnlock()
+		_ = inst.forAllStreams(func(s *stream) error {
+			streams = append(streams, s)
+			return nil
+		})
+		streamInstances[i] = streamInstance{
+			streams: streams,
+			id:      inst.instanceID,
 		}
-		close(ch)
-	}()
-	return ch
+	}
+	return &streamIterator{
+		instances: streamInstances,
+	}
+}
+
+// Next loads the next stream of the current instance.
+// If the instance is empty, it moves to the next instance until there is no more.
+// Return true if there's a next stream, each successful calls will replace the current stream.
+func (s *streamIterator) Next() bool {
+	if len(s.instances) == 0 {
+		s.instances = nil
+		return false
+	}
+	currentInstance := s.instances[0]
+	if len(currentInstance.streams) == 0 {
+		s.instances = s.instances[1:]
+		return s.Next()
+	}
+
+	// current stream
+	stream := currentInstance.streams[0]
+
+	// remove the first stream
+	s.instances[0].streams = s.instances[0].streams[1:]
+
+	stream.chunkMtx.RLock()
+	defer stream.chunkMtx.RUnlock()
+
+	if len(stream.chunks) < 1 {
+		// it's possible the stream has been flushed to storage
+		// in between starting the checkpointing process and
+		// checkpointing this stream.
+		return s.Next()
```

Comment:
I can't think of a deadlock scenario while holding the chunk mutex here and calling `s.Next()`, but I am worried about it.